quil-rs/src/program/scheduling/schedule.rs

```diff
@@ -0,0 +1,367 @@
+//! A Schedule represents a flattening of the [`DependencyGraph`] into a linear sequence of
+//! instructions, with each instruction assigned a start time and duration.
+
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use petgraph::{
+    visit::{EdgeFiltered, Topo},
+    Direction,
+};
+
+use crate::{
+    instruction::{
+        AttributeValue, Capture, Delay, Instruction, Pulse, RawCapture, WaveformInvocation,
+    },
+    quil::Quil,
+    Program,
+};
+
+use super::{ExecutionDependency, ScheduledBasicBlock, ScheduledGraphNode};
+
+#[derive(Clone, Default, PartialEq, PartialOrd)]
+pub struct Seconds(pub f64);
+
+impl std::ops::Add<Seconds> for Seconds {
+    type Output = Seconds;
+
+    fn add(self, rhs: Seconds) -> Self::Output {
+        Self(rhs.0 + self.0)
+    }
+}
+
+pub trait Zero {
+    fn zero() -> Self;
+}
+
+impl Zero for Seconds {
+    fn zero() -> Self {
+        Self(0.0)
+    }
+}
+
+// todo: rather than Default, it should require Time::zero. For primitives those are the same,
+// but not so for Expression
+#[derive(Debug)]
+pub struct Schedule<Time> {
+    items: Vec<ComputedScheduleItem<Time>>,
+    /// The total duration of the block. This is the end time of the schedule when it starts at Time::zero()
+    duration: Time,
+}
+
+impl<Time> Schedule<Time> {
+    pub fn duration(&self) -> &Time {
+        &self.duration
+    }
+
+    pub fn items(&self) -> &[ComputedScheduleItem<Time>] {
+        self.items.as_ref()
+    }
+}
+
+impl<Time: Zero> Default for Schedule<Time> {
+    fn default() -> Self {
+        Self {
+            items: Default::default(),
+            duration: Time::zero(),
+        }
+    }
+}
+
+pub type FixedSchedule = Schedule<Seconds>;
+
+#[derive(Debug)]
+pub struct ComputedScheduleItem<Time> {
+    pub time_span: TimeSpan<Time>,
+    pub instruction_index: usize,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ComputedScheduleError {
+    #[error("unknown duration for instruction {}", instruction.to_quil_or_debug())]
+    UnknownDuration { instruction: Instruction },
+
+    #[error("internal error: invalid dependency graph")]
+    InvalidDependencyGraph,
+}
+
+pub type ComputedScheduleResult<T> = Result<T, ComputedScheduleError>;
+
+/// Represents a span of time, for some unit of time
+#[derive(Debug)]
+pub struct TimeSpan<Time> {
+    /// The inclusive start time of the described item
+    start_time: Time,
+
+    /// The described item's continuous duration
+    duration: Time,
+}
+
+impl<Time> TimeSpan<Time> {
+    pub fn start_time(&self) -> &Time {
+        &self.start_time
+    }
+
+    pub fn duration(&self) -> &Time {
+        &self.duration
+    }
+}
+
+impl<'p> ScheduledBasicBlock<'p> {
+    // todo create more restricted subtype of instruction for ScheduledBasicBlock so we don't have
+    // to handle unreachable instructions here
```

Comment:
Reviewer thoughts? I tried out `subenum` but found it lacking:

* I'd like for it to allow the subenum to be a ref rather than taking ownership
* It doesn't allow for easy categorization, i.e. `InSubGroup` vs `NotInSubGroup`. It only has `InSubGroup::try_from(Parent)` which means that all variants of `Parent` must still be handled when `try_from` fails.

Reply:
It seems like, rather than a set of "sub-enums", what you're looking for might be a separate enum with fewer variants, and a mapping from `Instruction` to `SchedulableInstruction`, where each variant of `SchedulableInstruction` holds references to the relevant innards of `Instruction`?

Reply:
not quite - while that seems helpful in some ways, it makes the handling of instructions in, say, `impl<'p> From<&'p Program> for ControlFlowGraph<'p>` just as unwieldy.